onnxruntime/core/providers/webnn/builders/impl/lstm_op_builder.cc

```diff
@@ -0,0 +1,237 @@
+  const auto& input_defs = node.InputDefs();
+  if (input_defs.size() > 4 && input_defs[4]->Exists()) {
+    if (!Contains(initializers, input_defs[4]->Name())) {
+      LOGS(logger, ERROR) << "LSTM: sequence_lens must be constant and the values must be equal to input shape[0]";
+      return false;
+    }
+
+    std::vector<int64_t> input_shape;
+    if (!GetShape(*input_defs[0], input_shape, logger)) {
+      LOGS(logger, ERROR) << "Cannot get input's shape";
+      return false;
+    }
+    int32_t steps = static_cast<int32_t>(input_shape[0]);
+
+    const auto& sequence_lens_tensor = *initializers.at(input_defs[4]->Name());
+    std::vector<int32_t> sequence_lens;
+    if (!ReadIntArrayFrom1DTensor(sequence_lens_tensor, sequence_lens, logger)) {
+      LOGS(logger, ERROR) << "Cannot read sequence lens tensor";
+      return false;
+    }
+    if (!std::all_of(sequence_lens.begin(), sequence_lens.end(),
+                     [steps](int32_t lens) -> bool { return steps == lens; })) {
+      LOGS(logger, ERROR) << "LSTM: every sequence length must be equal to input shape[0]";
+      return false;
+    }
+  }
+
+  NodeAttrHelper helper(node);
+  if (helper.HasAttr("activations")) {
+    const auto activations = helper.Get("activations", std::vector<std::string>{"Sigmoid", "Tanh", "Tanh"});
+    if (activations.size() >= 6 &&
+        (activations[0] != activations[3] || activations[1] != activations[4] || activations[2] != activations[5])) {
+      LOGS(logger, ERROR) << "LSTM: forward and reverse directions must have the same activations";
+      return false;
+    }
+
+    if (activations.size() >= 6 && (helper.HasAttr("activation_alpha") || helper.HasAttr("activation_beta"))) {
+      LOGS(logger, ERROR) << "LSTM: activation_alpha and activation_beta are not supported when provided 6 activations";
+      return false;
+    }
+
+    if (!std::all_of(activations.begin(), activations.end(), [](const std::string& activation) -> bool {
+          return activation == "Affine" || activation == "Relu" || activation == "LeakyRelu" || activation == "Tanh" ||
+                 activation == "Sigmoid" || activation == "HardSigmoid" || activation == "Elu" ||
+                 activation == "Softsign" || activation == "Softplus";
+        })) {
+      LOGS(logger, ERROR) << "LSTM: activations must be one of Affine, Relu, LeakyRelu, Tanh, Sigmoid, HardSigmoid, "
+                             "Elu, Softsign, Softplus";
+      return false;
+    }
+  }
+
+  if (helper.Get("clip", std::numeric_limits<float>::max()) != std::numeric_limits<float>::max()) {
+    LOGS(logger, ERROR) << "LSTM: clip is not supported";
+    return false;
+  }
+
+  if (helper.Get("input_forget", 0) != 0) {
+    LOGS(logger, ERROR) << "LSTM: input_forget == 1 is not supported";
+    return false;
+  }
+
+  if (helper.Get("layout", 0) != 0) {
+    LOGS(logger, ERROR) << "LSTM: batchwise (layout == 1) is not supported";
```

Comment:
Not sure if it's achievable. I tried to run the fallback path in cpu with `layout == 1`, and it reports error:
> onnxruntime::LSTMBase::LSTMBase(const OpKernelInfo &) layout_ == 0 was false. Batchwise recurrent operations (layout == 1) are not supported.

cc @fdwr for more insights, thanks!

Reply:
You may find this comment in the DML EP relevant:

https://github.com/microsoft/onnxruntime/blob/main/onnxruntime/core/providers/dml/DmlExecutionProvider/src/Operators/DmlOperatorRecurrentNeuralNetwork.cpp#L409-L422

```
    // layout=1 for batchwise operation is unsupported, added in opset 14 for RNN, GRU, and LSTM
    // (https://github.com/onnx/onnx/pull/3217, https://github.com/onnx/onnx/pull/2284).
    // Currently (2022-05-27) the ORT CPU execution provider (lstm_base.h) does not support it either,
    // with no models warranting it. When needed, it can be achieved with no new DML API's by just
    // swapping the size and strides in the TensorDesc before filling in the *_OPERATOR_DESC, where:
    //
    // layout=0: (default, consistent with opset 7)
    //      X.shape = [seq_length, batch_size, input_size]
    //      Y.shape = [seq_length, num_directions, batch_size, hidden_size]
    //      initial_h.shape = Y_h.shape = initial_c.shape = Y_c.shape = [num_directions, batch_size, hidden_size]
    // layout=1:
    //      X.shape = [batch_size, seq_length, input_size]
    //      Y.shape = [batch_size, seq_length, num_directions, hidden_size]
    //      initial_h.shape = Y_h.shape = initial_c.shape = Y_c.shape = [batch_size, num_directions, hidden_size]
```